homeassistant/components/shelly/light.py

```diff
@@ -369,7 +367,14 @@ async def async_turn_on(self, **kwargs: Any) -> None:
                     self.wrapper.model,
                 )
 
-        if await self.set_light_mode(set_mode):
+        if (
+            set_mode
+            and self.mode != set_mode
+            and self.wrapper.model in DUAL_MODE_LIGHT_MODELS
+        ):
+            params["mode"] = set_mode
+            self.mode_result = self.control_result = await self.set_state(**params)
+        else:
             self.control_result = await self.set_state(**params)
 
```

Comment:
we don't need to check current mode now since now we know that the bulb supports mode change so we can just send the new mode together with other parameters, if the bulb is in the same mode it ignores the change.
Also no need for `mode_result` since we now use single API call
```suggestion
        if set_mode and self.wrapper.model in DUAL_MODE_LIGHT_MODELS:
            params["mode"] = set_mode

        self.control_result = await self.set_state(**params)
```